### PR TITLE
*: make  start scheduler more smoothly

### PR DESCRIPTION
**Pull request by nolouch**
close https://github.com/pingcap/pd/issues/625
PTAL @disksing  @andelf @siddontang

### CODE REVIEW

File: server/cache.go [hunk 1]
```diff
@@ -333,6 +333,7 @@ type clusterInfo struct {
 	stores  *storesInfo
 	regions *regionsInfo
 
+	reactiveRegions int
```
Comment (siddontang) [inline server/cache.go]: seem this will be increased all the time.

File: server/cache.go [hunk 2]
```diff
@@ -677,6 +684,9 @@ func (c *clusterInfo) handleRegionHeartbeat(region *RegionInfo) error {
 		}
 		if region.Leader.GetId() != origin.Leader.GetId() {
 			log.Infof("[region %d] Leader changed from {%v} to {%v}", region.GetId(), origin.GetPeer(origin.Leader.GetId()), region.GetPeer(region.Leader.GetId()))
+			if origin.Leader.GetId() == 0 {
+				c.reactiveRegions++
```
Comment (siddontang) [inline server/cache.go]: If PD restarts, it may load regions from etcd, seem the origin leader may be not 0.

File: server/cache.go [hunk 3]
```diff
@@ -624,6 +625,12 @@ func (c *clusterInfo) getFollowerStores(region *RegionInfo) []*storeInfo {
 	return stores
 }
 
+func (c *clusterInfo) isClusterInfoFullReported() bool {
```
Comment (siddontang) [inline server/cache.go]: should consider a better name

File: server/cache.go (hunk 1 above)
Comment (nolouch) [inline server/cache.go]: just the origin region load and leader change from nil will increase.

File: server/cache.go (hunk 2 above)
Comment (nolouch) [inline server/cache.go]: PD restart just loads meta region.

File: server/coordinator.go [hunk 4]
```diff
@@ -109,6 +111,17 @@ func (c *coordinator) dispatch(region *RegionInfo) *pdpb.RegionHeartbeatResponse
...
 
 func (c *coordinator) run() {
+	ticker := time.NewTicker(runSchedulerCheckInterval)
+	defer ticker.Stop()
+	for {
+		if c.shouldRun() {
+			break
+		}
+		select {
+		case <-ticker.C:
```
Comment (siddontang) [inline server/coordinator.go]: check context Done here

File: server/coordinator.go [hunk 5]
```diff
@@ -109,6 +111,17 @@ func (c *coordinator) dispatch(region *RegionInfo) *pdpb.RegionHeartbeatResponse
 }
 
 func (c *coordinator) run() {
+	ticker := time.NewTicker(runSchedulerCheckInterval)
+	defer ticker.Stop()
+	for {
```
Comment (siddontang) [inline server/coordinator.go]: add a log here

File: server/cache.go [hunk 6]
```diff
@@ -624,6 +625,13 @@ func (c *clusterInfo) getFollowerStores(region *RegionInfo) []*storeInfo {
 	return stores
 }
 
+// isPrepare check whether the clusterInfo is collected when the PD restart
+func (c *clusterInfo) isPrepare() bool {
```
Comment (siddontang) [inline server/cache.go]: isPrepared 

Btw, the comment mismatches.

File: server/balancer_test.go [hunk 7]
```diff
@@ -93,6 +93,17 @@ func (c *testClusterInfo) addLeaderRegion(regionID uint64, leaderID uint64, foll
 	c.putRegion(newRegionInfo(region, leader))
 }
 
+func (c *testClusterInfo) LoadRegion(regionID uint64, followerIds ...uint64) {
+	//  regions load from etcd will no leader
```
Comment (siddontang) [inline server/balancer_test.go]: will have

File: server/coordinator.go [hunk 8]
```diff
@@ -109,6 +111,21 @@ func (c *coordinator) dispatch(region *RegionInfo) *pdpb.RegionHeartbeatResponse
...
+		if c.shouldRun() {
+			break
+		}
+		log.Info("coordinator: Waiting for the cluster information to be collected")
```
Comment (siddontang) [inline server/coordinator.go]: what I mean is to add log before `for` and before `break`. 
No need to log here every time.

File: server/cache.go [hunk 9]
```diff
@@ -677,6 +685,10 @@ func (c *clusterInfo) handleRegionHeartbeat(region *RegionInfo) error {
 		}
 		if region.Leader.GetId() != origin.Leader.GetId() {
 			log.Infof("[region %d] Leader changed from {%v} to {%v}", region.GetId(), origin.GetPeer(origin.Leader.GetId()), region.GetPeer(region.Leader.GetId()))
+			log.Info("test %+v", origin.Leader)
```
Comment (siddontang) [inline server/cache.go]: remove this log

File: server/cluster.go [hunk 10]
```diff
@@ -86,7 +86,7 @@ func (c *RaftCluster) start() error {
 	c.cachedCluster = cluster
 
 	c.coordinator = newCoordinator(c.cachedCluster, c.s.scheduleOpt)
-	c.coordinator.run()
+	go c.coordinator.run()
```
Comment (siddontang) [inline server/cluster.go]: add 1 for the wait group before

File: server/coordinator_test.go [hunk 11]
```diff
@@ -211,6 +211,46 @@ func (s *testCoordinatorSuite) TestPeerState(c *C) {
...
+	co := newCoordinator(cluster, opt)
+
+	tc.LoadRegion(1, 1, 2, 3)
+	tc.LoadRegion(2, 1, 2, 3)
+	tc.LoadRegion(3, 1, 2, 3)
+	tc.LoadRegion(4, 1, 2, 3)
+	tc.LoadRegion(5, 1, 2, 3)
+	c.Assert(co.shouldRun(), IsFalse)
+
+	r1 := tc.getRegion(1)
```
Comment (siddontang) [inline server/coordinator_test.go]: should table-driven test

```go
tbl := []struct {
    RegionID int,
    shouldRun bool,
} {
    
}

for _, t := range tbl {
    r := tc.getRegion(t.RegionID)
    ...
}
```

**Comment by siddontang** [comment]
Rest LGTM

**Comment by siddontang** [comment]
PTAL @disksing

File: server/cache.go (hunk 1 above)
Comment (disksing) [inline server/cache.go]: reactive -> active?

**Comment by siddontang** [comment]
CI failed @nolouch

**Comment by nolouch** [comment]
the environmental reasons, report no leader.  I will merge it now

**Pull request merged**